crates/services/p2p/src/peer_report.rs

```diff
@@ -0,0 +1,452 @@
+const MAX_IDENTIFY_ADDRESSES: usize = 10;
+const HEALTH_CHECK_INTERVAL_IN_SECONDS: u64 = 10;
+const REPUTATION_DECAY_INTERVAL_IN_SECONDS: u64 = 1;
+
+/// Events emitted by PeerInfoBehaviour
+#[derive(Debug, Clone)]
+pub enum PeerReportEvent {
+    PeerConnected {
+        peer_id: PeerId,
+        addresses: Vec<Multiaddr>,
+        initial_connection: bool,
+    },
+    PeerDisconnected {
+        peer_id: PeerId,
+    },
+    PeerIdentified {
+        peer_id: PeerId,
+        agent_version: String,
+        addresses: Vec<Multiaddr>,
+    },
+    PeerInfoUpdated {
+        peer_id: PeerId,
+        block_height: BlockHeight,
+    },
+    /// Informs p2p service / PeerManager to check health of reserved nodes' connections
+    CheckReservedNodesHealth,
+    /// Informs p2p service / PeerManager to perform reputation decay of connected nodes
+    PerformDecay,
+}
+
+// `Behaviour` that reports events about peers
+pub struct PeerReportBehaviour {
+    heartbeat: Heartbeat,
+    identify: Identify,
+    pending_events: VecDeque<PeerReportEvent>,
+    // regulary checks if reserved nodes are connected
+    health_check: Interval,
+    decay_interval: Interval,
+}
+
+impl PeerReportBehaviour {
+    pub(crate) fn new(config: &Config) -> Self {
+        let identify = {
+            let identify_config =
+                IdentifyConfig::new("/fuel/1.0".to_string(), config.keypair.public());
+            if let Some(interval) = config.identify_interval {
+                Identify::new(identify_config.with_interval(interval))
+            } else {
+                Identify::new(identify_config)
+            }
+        };
+
+        let heartbeat =
+            Heartbeat::new(config.heartbeat_config.clone(), BlockHeight::default());
+
+        Self {
+            heartbeat,
+            identify,
+            pending_events: VecDeque::default(),
+            health_check: time::interval(Duration::from_secs(
+                HEALTH_CHECK_INTERVAL_IN_SECONDS,
+            )),
+            decay_interval: time::interval(Duration::from_secs(
+                REPUTATION_DECAY_INTERVAL_IN_SECONDS,
+            )),
+        }
+    }
+
+    pub fn update_block_height(&mut self, block_height: BlockHeight) {
+        self.heartbeat.update_block_height(block_height);
+    }
+}
+
+impl NetworkBehaviour for PeerReportBehaviour {
+    type ConnectionHandler = IntoConnectionHandlerSelect<
+        <Heartbeat as NetworkBehaviour>::ConnectionHandler,
+        <Identify as NetworkBehaviour>::ConnectionHandler,
+    >;
+    type OutEvent = PeerReportEvent;
+
+    fn new_handler(&mut self) -> Self::ConnectionHandler {
+        IntoConnectionHandler::select(
+            self.heartbeat.new_handler(),
+            self.identify.new_handler(),
+        )
+    }
+
+    fn addresses_of_peer(&mut self, peer_id: &PeerId) -> Vec<Multiaddr> {
+        self.identify.addresses_of_peer(peer_id)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm<Self::ConnectionHandler>) {
```

Comment:
Cool, now this method and the file itself contain only propagation events logic=)

I don't know when `on_swarm_event` is called. Maybe in the future, we also will be able to move the handling of some events directly into `PeerManager`, but it is already good for now=)

Reply:
Yes things are much nicer in `PeerReport`.

> Maybe in the future, we also will be able to move the handling of some events directly into `PeerManager`, but it is already good for now=)

If we did this we would end up with the previous issue we just had, where `PeerManager` would become a Behaviour and in order to talk to p2p_service it would need to constantly push events up to the p2p_service (ie `PeerManagerEvent::BanPeer(peerId)`)

`on_swarm_event` comes from the swarm -> behaviour -> p2p_service

Reply:
It does seem strange to have all this logic unrelated to reputation included here (e.g. notifying swarm about incoming connections). I'm fine with it for now though if we open up a ticket to revisit this.